cobc/replace.c

```diff
@@ -0,0 +1,841 @@
+			       const struct cb_replace_list *replace_list);
+
+static struct cb_token_list *
+token_list_add (WITH_DEPTH struct cb_token_list *list,
+		const char *text,
+		const char *token);
+
+/* This specific token_list_add function does a standard append on
+   list, without expecting `last` field to be correctly set.  This is
+   important as `pp_token_list_add` only correctly works when always
+   adding on the same head, other `last` fields in the middle of the
+   list not being correctly updated...
+ */
+static
+struct cb_token_list *
+token_list_add (WITH_DEPTH struct cb_token_list *list,
+		const char *text, const char *token)
+{
+#ifdef DEBUG_REPLACE_TRACE
+	fprintf(stderr, "%stoken_list_add(%s,'%s')\n",
+		DEPTH, string_of_list(list), text);
+#endif
+	struct cb_token_list	*p;
+
+	p = cobc_plex_malloc (sizeof (struct cb_token_list));
+	p->text = cobc_plex_strdup (text);
+	if (token == NULL) {
+		p->token = NULL;
+	} else {
+		p->token = cobc_plex_strdup (token);
+	}
+
+	p->next = NULL;
+	if (list==NULL) {
+		return p;
+	} else {
+		struct cb_token_list *cursor = list;
+		for(;cursor->next != NULL; cursor = cursor->next);
+		cursor->next = p;
+		return list;
+	}
+}
+
+
+
+static
+const void pop_token (WITH_DEPTH struct cb_replacement_state *repls,
+		      const char **text, const char **token)
+{
+	const struct cb_token_list *q = repls->token_queue ;
+	repls->token_queue = q->next ;
+#ifdef DEBUG_REPLACE_TRACE
+	fprintf(stderr, "%spop_token(%s) -> '%s'\n",
+		DEPTH, repls->name, q->text);
+#endif
+	if (text) *text = q->text ;
+	if (token) *token = q->token ;
+}
+
+static
+void ppecho_switch (WITH_DEPTH struct cb_replacement_state *repls,
+		    const char* text, const char* token)
+{
+#ifdef DEBUG_REPLACE_TRACE
+	fprintf(stderr, "%sppecho_switch(%s, '%s')\n",
+		DEPTH, repls->name, text);
+#endif
+	switch( repls->ppecho ){
+	case CB_PPECHO_DIRECT:
+#ifdef DEBUG_REPLACE
+		fprintf(stderr, "%s ppecho_direct('%s')\n", DEPTH, text);
+#endif
+		return cb_ppecho_direct (text, token);
+	case CB_PPECHO_REPLACE:
+		return ppecho_replace (MORE_DEPTH text, token);
+	}
+}
+
+static
+void ppecho_switch_text_list (WITH_DEPTH struct cb_replacement_state *repls,
+			 const struct cb_text_list *p)
+{
+#ifdef DEBUG_REPLACE_TRACE
+	fprintf(stderr, "%sppecho_switch_text_list(%s, %s)\n",
+		DEPTH, repls->name, string_of_list(p));
+#endif
+
+	for (;p;p=p->next){
+		ppecho_switch (MORE_DEPTH repls, p->text, NULL);
+	}
+}
+
+
+static
+void ppecho_switch_token_list (WITH_DEPTH struct cb_replacement_state *repls,
+			 const struct cb_token_list *p)
+{
+#ifdef DEBUG_REPLACE_TRACE
+	fprintf(stderr, "%sppecho_switch_token_list(%s, %s)\n",
+		DEPTH, repls->name, string_of_list(p));
+#endif
+
+	for (;p;p=p->next){
+		ppecho_switch (MORE_DEPTH repls, p->text, p->token);
+	}
+}
+
+static
+int is_leading_or_trailing (WITH_DEPTH int leading,
+			    const char* src_text,
+			    const char* text,
+			    int strict)
+{
+
+	size_t src_len = strlen (src_text);
```

Comment:
The length  variables can be const, no? Do we need to strlen on each access of these or do we have the size in the tructs from the caller?